REDESIGN.md

```diff
@@ -0,0 +1,143 @@
+  * Add acceptance tests for any missing service classes
+    * [foreman](https://github.com/theforeman/puppet-certs/pull/355)
+    * [qpid_router](https://github.com/theforeman/puppet-certs/pull/356)
+    * [pulp_client](https://github.com/theforeman/puppet-certs/pull/357)
+    * [puppet](https://github.com/theforeman/puppet-certs/pull/358)
+  * Add certificates from /root/ssl-build/${hostname} into tarball creation
+    * Needed for being able to deploy certificates from the build directory on a content proxy as the current design only copies over RPMs and then installs those RPMs resulting in the `/etc/pki/katello-certs-tools` location being available
+    * [Pull Request](https://github.com/theforeman/puppet-certs/pull/352)
+  * Investigate configuring where certificates are generated for a given service by katello-ssl-tool
+  * Add new type for managing a certificate
+    * This type and provider would handle copying the provided certificate to a destination, validating it and setting properties on the file
+    * Name ideas:
+      * managed_certificate
+      * certificate
+      * certificate_file
+      * public_certificate
+      * service_certificate
+  * Add a new provider for managing a private key
+    * This type and provider would handle copying the provided certificate to a destination, validating it and setting properties on the file
+    * Name ideas:
+      * managed_private_key
+      * private_key
+      * private_key_file
+      * service_private_key
+  * Add a new defines `key_pair` that provides a single interface to wrap private and public key management
+    * This will eventually replace `keypair` and provides a way to migrate service by service in smaller chunks
+  * Switch classes one by one to deploy certificates from the build directory (e.g. /root/ssl-build) rather than from the RPM location
+    * Set `deploy => false` on `cert` provider to avoid RPM install
+  * Remove RPM deployment and generation
+    * Add a change that stops generating RPMs when `deploy => false` is set on the `cert` provider
+    * Clean out RPM handling code from `cert` provider
+
+### Deploy and Manage certificates within service modules
+
+#### Problem Statement
+
+Services almost always need to set some level of ownership on certificates in order to consume them at runtime. This can be as simple as user and group. The certificates used by a service are also often co-located within a directory owned by that service. This creates dependency cycles and burdens when trying to manage the certificates from within puppet-certs. For example, puppet-candlepin needs to install Candlepinand ensure the tomcat user and group exists, as well as the `/etc/candlepin/certs` directory in order for puppet-certs to set the owner and group on the certificates. And, puppet-candlepin needs puppet-certs to have created and deployed the certificates with those correct ownerships before the service can be started.
```

Comment:
```suggestion
Services almost always need to set some level of ownership on certificates in order to consume them at runtime. This can be as simple as user and group. The certificates used by a service are also often co-located within a directory owned by that service. This creates dependency cycles and burdens when trying to manage the certificates from within puppet-certs. For example, puppet-candlepin needs to install Candlepin and ensure the tomcat user and group exists, as well as the `/etc/candlepin/certs` directory in order for puppet-certs to set the owner and group on the certificates. And, puppet-candlepin needs puppet-certs to have created and deployed the certificates with those correct ownerships before the service can be started.
```